clang/include/clang/AST/PreorderAST.h

```diff
@@ -0,0 +1,174 @@
+//===------- PreorderAST.h: An n-ary preorder abstract syntax tree -------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the interface for an n-ary preorder abstract syntax tree
+//  which is used to semantically compare two expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_PREORDER_AST_H
+#define LLVM_CLANG_PREORDER_AST_H
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/CanonBounds.h"
+#include "clang/AST/Expr.h"
+
+namespace clang {
+  // Each node of the PreorderAST has a variable-count list which stores the
+  // name of each variable along with a count of the number of times the
+  // variable appears in the node.
+  struct VarTy {
+    std::string name;
+    llvm::APSInt count;
+  };
+
+  // Each node of the PreorderAST contains 3 fields:
+  // - opcode: The opcode for the node.
+  // - variable-count list: A list of variables and their counts.
+  // - constant value: All constants occurring in the node are constant folded.
+
+  // For example, consider an expression E = a + 3 + b + 4 + a
+  // A node of the PreorderAST for E would contain the following:
+  // - opcode: +
+  // - variable-count list: [a:2, b:1]
+  // - constant value: 3 + 4 = 7
+
+  using VarListTy = std::vector<VarTy>;
+  using OpcodeTy = BinaryOperator::Opcode;
+  using ConstTy = llvm::APSInt;
+  using Result = Lexicographic::Result;
+
+  class PreorderAST {
+    class ASTNode {
+    public:
+      ASTContext &Ctx;
+      OpcodeTy opcode;
```

Comment:
Is there a precedent for using camelCase names elsewhere in clang? Other files such as SemaBounds, BoundsAnalysis, CanonBounds, etc. seem to use PascalCase.